src/System.Private.CoreLib/src/System/Threading/ClrThreadPool.WaitThread.cs

```diff
@@ -0,0 +1,425 @@
+                        {
+                            if (_pendingRemoves[i] == _registeredWaits[j])
+                            {
+                                _registeredWaits[j] = _registeredWaits[_numUserWaits - 1];
+                                _waitHandles[j + 1] = _waitHandles[_numUserWaits];
+                                _registeredWaits[_numUserWaits - 1] = null;
+                                _waitHandles[_numUserWaits] = null;
+                                for (int k = _numUserWaits - 1; k > 0; k--)
+                                {
+                                    if (_registeredWaits[k - 1] != null)
+                                    {
+                                        _numUserWaits = k;
+                                        break;
+                                    }
+                                }
+
+                                if(_registeredWaits[0] == null)
+                                {
+                                    _numUserWaits = 0;
+                                }
+
+                                break;
+                            }
+                        }
+                        _pendingRemoves[i] = null;
+                    }
+                    _numPendingRemoves = 0;
+
+                    Debug.Assert(originalNumUserWaits - originalNumPendingRemoves == _numUserWaits,
+                        $"{originalNumUserWaits} - {originalNumPendingRemoves} == {_numUserWaits}");
+                }
+                finally
+                {
+                    _registeredHandlesLock.Release();
+                    _removesLock.Release();
+                }
+            }
+            
+            /// <summary>
+            /// Queue a call to <see cref="CompleteWait(object)"/> on the ThreadPool.
+            /// </summary>
+            /// <param name="registeredHandle">The handle that completed.</param>
+            /// <param name="timedOut">Whether or not the wait timed out.</param>
+            private void QueueWaitCompletion(RegisteredWait registeredHandle, bool timedOut)
+            {
+                ThreadPool.QueueUserWorkItem(CompleteWait, new CompletedWaitHandle(registeredHandle, timedOut));
+            }
+
+            /// <summary>
+            /// Process the completion of a user-registered wait.
+            /// </summary>
+            /// <param name="state">A <see cref="CompletedWaitHandle"/> object representing the wait completion.</param>
+            private void CompleteWait(object state)
+            {
+                CompletedWaitHandle handle = (CompletedWaitHandle)state;
+                handle.CompletedHandle.CanUnregister.Reset();
+                handle.CompletedHandle.PerformCallback(handle.TimedOut);
+                handle.CompletedHandle.CanUnregister.Set();
+                if (!handle.CompletedHandle.Repeating)
+                {
+                    QueueOrExecuteUnregisterWait(handle.CompletedHandle);
+                }
```

Comment:
With the current design this is not possible. The unregistration would cause `PerformCallback` to not call the callback. I'll look into trying to redesign so we can make this common path quicker.